### PR TITLE
Bluetooth: Host: Find by type should accept 128bit UUIDs

### PR DESCRIPTION
**Pull request by kimsekkelund**
Find by type does only accept a UUID with the same length as the UUID
which is stored in the internal list. If a UUID is stored in the short
16 bit format then a request with 128 bit UUID will fail.
Add support for the missing formats.

Signed-off-by: Kim Sekkelund <ksek@oticon.com>

### CODE REVIEW

File: subsys/bluetooth/host/att.c
Comment (Vudentz) [inline subsys/bluetooth/host/att.c]: While this seems correct I wonder if it wouldn't be better to create a helper function to convert from over the air format to struct bt_uuid, i.e. bt_uuid_create(&uuid, size, data), which basically consolidates the code above which is already duplicated in a few places include uuid_create that that uses a net_buf.

Comment (kimsekkelund) [reply subsys/bluetooth/host/att.c]: I agree. I will work on a change for that.
The 32-bit UUID is not allowed over the air, so should I remove that case?
Core5.1: Vol3, partF, sect3: "All 32-bit Attribute UUIDs shall be converted to 128-bit UUIDs when the Attribute UUID is contained in an ATT PDU."

Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: > I agree. I will work on a change for that.
> The 32-bit UUID is not allowed over the air, so should I remove that case?
> Core5.1: Vol3, partF, sect3: "All 32-bit Attribute UUIDs shall be converted to 128-bit UUIDs when the Attribute UUID is contained in an ATT PDU."

Id leave 32bit since the uuid.c has already handling for it, we can have a check before calling the function to check if the size is valid for the bearer, or alternatively we a le variant, bt_uuid_le_create which does the check before calling bt_uuid_create.

Comment (kimsekkelund) [reply subsys/bluetooth/host/att.c]: Is there a bug in uuid_create()? It uses net_buf_pull_le16() for the 16-bit UUID but memcpy instead of net_buf_pull_mem() for the 128-bit UUID.
If it is okay NOT to decrease buf->len and increase buf->data, in uuid_create(), then we can call the new bt_uuid_create, otherwise we can not benefit from reusing code (other than make it easier to read).

Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: You can always do it advance the buf data manually, regarding the reason we use memcpy is that net_buf_pull_mem don't really copy anything to bt_uuid_128 to be able to do things like bt_uuid_cmp, it would perhaps be possible to just reference with pointer instead of having val[16] but we would probably add 4 bytes extra per UUID 128 bits so I rather leave like it is and do the memcpy.

File: subsys/bluetooth/host/att.c
Comment (kimsekkelund) [reply subsys/bluetooth/host/att.c]: @joerchan , I can not tell if it is better - the union method is used in many other places, so I figured that it is an accepted way to do it - and it also pass the sanity check :-)

File: subsys/bluetooth/host/att.c
Comment (kimsekkelund) [reply subsys/bluetooth/host/att.c]: @joerchan , the attr->read() function reads from the internal list of characteristics in CPU endianness where bt_uuid_create() reads from the receivied packet which is in little endian format.

File: subsys/bluetooth/host/att.c
Comment (joerchan) [reply subsys/bluetooth/host/att.c]: It's done with the union when used for storage, not when it is a pointer.
Also bt_uuid_* functions are in uuid.c. If we are keeping the name we should move it in there.

Comment (kimsekkelund) [reply subsys/bluetooth/host/att.c]: OK, I will move the function to uuid.c, and use the BT_UUID_16 macro instead.

File: subsys/bluetooth/host/att.c
Comment (joerchan) [reply subsys/bluetooth/host/att.c]: I understand. Would it perhaps be cleaner to create a different function for that as well? And document when they both can be used?
Also the uuid_create (net_buf version) could also re-use bt_uuid_create by using the req->uuid. 
IMO it would make this cleaner to have one function to create uuid from the received packet.

Comment (kimsekkelund) [reply subsys/bluetooth/host/att.c]: @joerchan , I agree that it would be cleaner, and have also considered to create a separate function, but I stopped because it would be a function that seems to be only related to the attribute read callback. Would we use the functionality at other places, or can I just put it in att.c as a local function?
With regard to the uuid_create(), I am on thin ice. I am a little bit afraid to break code by not doing proper buf->len and buf->data manipulation. But I can give it a try, and then the reviewers can vote.

Comment (joerchan) [reply subsys/bluetooth/host/att.c]: I think we should make them equally available, so perhaps `bt_uuid_create`, and `bt_uuid_create_le` the latter one creating uuid from Little Endian data format.

Regarding uuid_create(), as you pointed out here: https://github.com/zephyrproject-rtos/zephyr/pull/17765#discussion_r307730536 the buf->len is inconsistely moved depending on the UUID type. To me it seems like the reason this is not a problem is because the buf is not used after, otherwise I don't understand how that could not be an issue.
It seems like this is not the intention, and could lead to problems in the future. That is a very unexpected behaviour of the function IMO.

File: include/bluetooth/uuid.h
Comment (kimsekkelund) [inline include/bluetooth/uuid.h]: I missed description of the return value. Fixing that in a new PR

File: subsys/bluetooth/host/att.c
Comment (kimsekkelund) [reply subsys/bluetooth/host/att.c]: @joerchan , I have replaced uuid_create() with my bt_uuid_create() function, and it seems to work (at least for my test case).

File: include/bluetooth/uuid.h
Comment (joerchan) [inline include/bluetooth/uuid.h]: I would suggest something like this:
` * @return true if the data was valid and the UUID was successfully created.`

File: include/bluetooth/uuid.h
Comment (joerchan) [inline include/bluetooth/uuid.h]: I would suggest something like this:
`* @return true if the data was valid and the UUID was successfully created.`

File: subsys/bluetooth/host/att.c
Comment (joerchan) [inline subsys/bluetooth/host/att.c]: Return value not checked. Do we need to handle this?

**Review by joerchan — CHANGES_REQUESTED** [review]
This is looking very good now. Just a few minor comments.

File: subsys/bluetooth/host/att.c
Comment (kimsekkelund) [reply subsys/bluetooth/host/att.c]: Checks are added in next commit

File: include/bluetooth/uuid.h
Comment (kimsekkelund) [reply include/bluetooth/uuid.h]: Descriptions will be changed to the suggested in next commit

**Review by joerchan — APPROVED** [review]
Looks good. Thank you.

File: subsys/bluetooth/host/att.c
Comment (Vudentz) [inline subsys/bluetooth/host/att.c]: I rather bail out if bt_uuid_create_le fails, either with a goto or a return.

File: subsys/bluetooth/host/uuid.c
Comment (Vudentz) [inline subsys/bluetooth/host/uuid.c]: Add a empty line after declarations.

Comment (kimsekkelund) [reply subsys/bluetooth/host/uuid.c]: Added in next commit

File: subsys/bluetooth/host/att.c
Comment (kimsekkelund) [reply subsys/bluetooth/host/att.c]: Will be replaced by gotos in next commit

File: subsys/bluetooth/host/att.c
Comment (Vudentz) [inline subsys/bluetooth/host/att.c]: Call it just skip.

File: subsys/bluetooth/host/att.c
Comment (Vudentz) [inline subsys/bluetooth/host/att.c]: Perhaps we should log something if we fail to create the UUID, i.e. BT_WARN("Unable to create UUID: size %d", read)...

**Review by Vudentz — APPROVED** [review]
LGTM

**Comment by ioannisg** [comment]
Does this needs to be back-ported to 1.14?

**Comment by aescolar** [comment]
> Does this needs to be back-ported to 1.14?

@ioannisg This is a new feature. So I would say no.

**Comment by aescolar** [comment]
This has 2 approvals from core maintainers of this area and codeowners, but @jhedberg won't be here until after 2.0 is frozen. So merging so it can be into 2.0.

**Pull request merged**
